apiserver/facades/agent/machine/machiner.go

```diff
@@ -353,14 +335,49 @@ func (api *MachinerAPI) RecordAgentStartInformation(ctx context.Context, args pa
 	}
 
 	for i, arg := range args.Args {
-		m, err := api.getMachine(arg.Tag, canModify)
-		if err != nil {
-			results.Results[i].Error = apiservererrors.ServerError(err)
-			continue
-		}
-		if err := m.RecordAgentStartInformation(arg.Hostname); err != nil {
-			results.Results[i].Error = apiservererrors.ServerError(err)
-		}
+		err := api.recordAgentStartInformation(ctx, arg.Tag, arg.Hostname, canModify)
+		results.Results[i].Error = apiservererrors.ServerError(err)
 	}
 	return results, nil
 }
+
+func (api *MachinerAPI) recordAgentStartInformation(ctx context.Context, tag, hostname string, authChecker common.AuthFunc) error {
+	mTag, err := api.canModify(tag, authChecker)
+	if err != nil {
+		return err
+	}
+
+	mUUID, err := api.machineService.GetMachineUUID(ctx, machine.Name(mTag.Id()))
+	if errors.Is(err, machineerrors.MachineNotFound) {
+		return errors.NotFoundf("machine %q", mTag.Id())
+	} else if err != nil {
+		return err
+	}
+
+	return api.machineService.SetMachineHostname(ctx, mUUID, hostname)
```

Comment:
I think we need another 
```
if errors.Is(err, machineerrors.MachineNotFound) {
	return errors.NotFoundf("machine %q", mTag.Id())
} else if err != nil {
	return err
}
```
here in case something changes between the transactions